opsimulator/opsimulator.go

```diff
@@ -216,9 +247,143 @@ func (opSim *OpSimulator) AddDependency(chainID uint64, depSet []uint64) error {
+			return fmt.Errorf("failed to decode executing messages from transaction logs: %w", err)
+		}
+
+		if executingMessage != nil {
+			executingMessages = append(executingMessages, *executingMessage)
+		}
+	}
+
+	if len(executingMessages) >= 1 {
+		for _, executingMessage := range executingMessages {
+			client := opSim.clients[executingMessage.Identifier.ChainId.Uint64()]
+			if client == nil {
+				return fmt.Errorf("no client found for chainid: %v on executing message", executingMessage.Identifier.ChainId)
+			}
+
+			block, err := client.EthBlockByNumber(ctx, executingMessage.Identifier.BlockNumber)
+			if err != nil {
+				return fmt.Errorf("failed to fetch executing message block: %w", err)
+			}
+
+			if executingMessage.Identifier.Timestamp.Cmp(new(big.Int).SetUint64(block.Time())) != 0 {
+				return fmt.Errorf("executing message identifier does not match block timestamp: %w", err)
+			}
+
+			logs, err := client.EthGetLogs(
+				ctx,
+				ethereum.FilterQuery{
+					Addresses: []common.Address{executingMessage.Identifier.Origin},
+					FromBlock: executingMessage.Identifier.BlockNumber,
+					ToBlock:   executingMessage.Identifier.BlockNumber,
+				},
+			)
+			if err != nil {
+				return fmt.Errorf("failed to fetch initiating message logs: %w", err)
+			}
+			var initiatingMessageLogs []ethTypes.Log
+			for _, log := range logs {
+				logIndex := big.NewInt(int64(log.Index))
+				if logIndex.Cmp(executingMessage.Identifier.LogIndex) == 0 {
+					initiatingMessageLogs = append(initiatingMessageLogs, log)
+				}
+			}
+
+			if len(initiatingMessageLogs) == 0 {
+				return fmt.Errorf("initiating message not found")
+			}
+
+			// Since we look for a log at a specific index, this should never be more than 1.
+			if len(initiatingMessageLogs) > 1 {
+				return fmt.Errorf("unexpected number of initiating messages found: %v", len(initiatingMessageLogs))
+			}
+
+			if !bytes.Equal(executingMessage.Message, messagePayloadBytes(&initiatingMessageLogs[0])) {
+				return fmt.Errorf("executing and initiating message fields are not equal")
+			}
+		}
+	}
+
+	return nil
+}
+
+func toSimulatedLogs(call config.TraceCallRaw) []ethTypes.Log {
+	var logs []ethTypes.Log
+
+	// Recursively append the logs for each nested call
+	for _, call := range call.Calls {
+		logs = append(logs, toSimulatedLogs(call)...)
+	}
+
+	for _, log := range call.Logs {
+		logs = append(logs, ethTypes.Log{
+			Address: log.Address,
+			Topics:  log.Topics,
+			Data:    common.FromHex(log.Data),
+		})
+	}
+
+	return logs
+}
+
+func (opSim *OpSimulator) decodeRawTransaction(rawTx string) (*ethTypes.Transaction, error) {
+	rawTxBytes, err := hex.DecodeString(rawTx[2:])
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode raw transaction: %w", err)
+	}
+
+	var ethTx ethTypes.Transaction
+	if err := ethTx.UnmarshalBinary(rawTxBytes); err != nil {
+		// If UnmarshalBinary failed, we will try RLP in case user made mistake
+		e := rlp.DecodeBytes(rawTxBytes, &ethTx)
```

Comment:
In what case in particular does this catch? the binary serialization is RLP but generally with a single type byte in front, but `UnmarshalBinary` should handle both cases